// react-components/src/LoadingButton/index.js
import React, { Component } from 'react';
import ReactLoading from 'react-loading';

export default class LoadingButton extends Component {

  render() {

    const {
      name = 'button',
      type = 'submit',
      label,
      loading,
      onClick,
      buttonStyle = {},
      buttonClassNames
    } = this.props;

    const classNameStr = buttonClassNames || "button blue wide cropBottom";

    const loader = (
      <div style={styles.loaderContainer}>
        <ReactLoading
          type='spin'
          color='#2D8EFF'
          height={40}
          width={40}
        />
      </div>
    )

    const button = (
      <button
        type={type}
        onClick={onClick}
        className={classNameStr}
        style={{
          ...styles.button,
          ...buttonStyle
        }}
        name={name}
      >
        {label}
      </button>
    );

    const display = loading ? loader : button;

    return (
      <div>
        {display}
      </div>
    )
  }
}

const styles = {
  loaderContainer: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: 52,
    backgroundColor: '#FFF',
  },
  button: {
    width: 170,
    height: 52,
    fontSize: 16,
    fontWeight: '700',
    textAlign: 'center',
    fontFamily: 'proxima-nova',
  },
  buttonMobile: {
    width: 300,
    fontSize: 16,
    fontWeight: '700',
    height: 52,
    textAlign: 'center',
  },
}
