// react-components/src/AuthCode/index.js
import React, { Component } from 'react';
import { SlideDown } from 'react-slidedown';
import 'react-slidedown/lib/slidedown.css';

import { styles } from '../../utils/styles';
import Container from '../Shared/Container';
import BottomLinks from '../Shared/BottomLinks';

export default class AuthCode extends Component {

  constructor(props) {
    super(props);

    this.state = {
      verification_code: '',
      error: false,
      errorMessage: '',
    };

    this.onClick = this.onClick.bind(this);
    this.updateVerificationCode = this.updateVerificationCode.bind(this);
  }


  onClick(event) {
    event.preventDefault();

  }

  updateVerificationCode(event) {
    this.setState({
      verification_code: event.target.value,
    });
  }

  render () {
    const { verification_code, error, errorMessage } = this.state;
    const { customStyles = {}, displayComponent } = this.props;
    const { containerStyle = {}, errorStyle = {}, formStyle = {}, labelContainerStyle = {}, labelStyle = {}, inputStyle = {}, buttonStyle = {}, linkStyle = {}} = customStyles;


    return (
      <div>
        <Container>
          <div style={{
            display: 'flex',
            flexDirection: 'column',
            textAlign: 'left',
          }}>
          <h2
            style={{
              ...styles.header,
            }}
          >
            Verify your login
          </h2>

          <p style={styles.statement}>
            Please enter the 6-digit verification code from your Authenticator app.
          </p>

          <div
            style={{
              ...styles.labelContainer,
              ...labelContainerStyle,
            }}
          >
            <label
              style={{
                ...styles.label,
                ...labelStyle
              }}
              htmlFor="verification_code"
            >
              Enter Code
            </label>
          </div>
          <input
            style={{
              ...styles.input,
              ...inputStyle,
            }}
            type="text"
            name="verification_code"
            placeholder="Verification Code"
            value={verification_code}
            onChange={this.updateVerificationCode}
            required
          />
          <SlideDown
            closed={!error}
          >
            <p
              style={{
                ...styles.error,
                ...errorStyle
              }}
            >
              { errorMessage }
            </p>
          </SlideDown>

          <br/>
          <div
            style={{
              ...styles.bottomLine
            }}
          >
            <span
            style={{
              ...styles.statement
            }}
            >
              <a
                style={{
                  ...styles.link
                }}
                href="#"
                onClick={() => displayComponent('login')}
              >
                Cancel
              </a>
            </span>
            <button
              style={{
                ...styles.button
              }}
              onClick={this.onClick}
            >
              Verify
            </button>
          </div>

          </div>
        </Container>
        <BottomLinks />
      </div>
    );
  }
}
