// react-components/src/Shared/Banner.js
import React from 'react';

import PresearchLogo from './PresearchLogo';

const Banner = ({ containerStyle = {}, logoStyle = {}, headerStyle = {} }) => {
  return (
    <div
      style={{
        ...styles.container,
        ...containerStyle
      }}
    >
      <PresearchLogo
        fill='#FFFFFF'
        scale={.8}
        alt='Presearch'
        style={{
          ...styles.logo,
          ...logoStyle
        }}
      />
      <span
        style={{
          ...styles.header,
          ...headerStyle
        }}>
        Presearch
      </span>
    </div>
  );
}

const styles = {
  container: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center',
    backgroundColor: '#2D8EFF',
    paddingLeft: '40px',
    borderTopLeftRadius: '3px',
    borderTopRightRadius: '3px',
    border: '1px solid #2D8EFF',
    width: '100%',
    height: '100px',
  },
  logo: {
    backgroundColor: '#2D8EFF',
    width: '37px',
    height: '37px',
  },
  header: {
    fontSize: '30px',
    color: '#FFFFFF',
    marginLeft: '15px',
  }
};

export default Banner;
