// react-components/src/Shared/Container.js
import React from 'react';

import Banner from './Banner';
import Typekit from 'react-typekit';

const Container = (props) => {

  return (
    <div
      style={{
        ...styles.container,
        ...props.containerStyle,
      }}
    >
      <Typekit kitId="kfo7hwc" />
      <Banner />
      <div
        style={{
          ...styles.content,
          ...props.contentStyle,
        }}
      >
        <h2
          style={{
            ...styles.header,
          }}
        >
          { props.contentTitle}
        </h2>
        {props.children}
      </div>
    </div>
  );
}

const styles = {
  container: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'center',
    width: '450px',
    minHeight: '450px',
    fontFamily: 'proxima-nova',
  },
  content: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
    border: '1px solid #E5E5E5',
    backgroundColor: '#FFFFFF',
    padding: '35px',
    width: '100%',
    height: '100%',
    minHeight: '350px',
    borderBottomLeftRadius: '3px',
    borderBottomRightRadius: '3px',
    borderTop: 'transparent',
  },
  header: {
    fontSize: '22px',
    fontFamily: 'Arial',
    fontWeight: '300',
    lineHeight: '27px',
    color: '#494949',
    paddingBottom: 20,
  },
};

export default Container;
