// react-components/src/Shared/RememberMe.js
import React, { Component } from 'react';

import Cookies from 'universal-cookie';

export default class RememberMe extends Component {
  constructor(props) {
    super(props);

    this.state = {
      remember_me: false,
    };

    this.onChange = this.onChange.bind(this);
  }

  componentDidMount() {
    const cookies = new Cookies();
    const remember_me = cookies.get('remember_me');

    if (remember_me && (remember_me === 'true' || remember_me === true)) {
      this.setState({
        remember_me: true,
      });
    }
  }

  onChange(event) {
    const target = event.target;
    const cookies = new Cookies();

    cookies.set('remember_me', !this.state.remember_me,  { path: '/' });

    this.setState({
      remember_me: !this.state.remember_me,
    });
  }


  render() {
    const { remember_me } = this.state;

    return (
      <div style={styles.container}>
        <input
          style={{
            ...styles.checkbox,
          }}
          type="checkbox"
          checked={remember_me}
          onChange={this.onChange}
        />
        <span
          style={{
            ...styles.rememberMe
          }}
        >
          Remember me on this computer.
        </span>
      </div>
    );
  }
}

const styles = {
  container: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center',
    width: '100%',
    height: 22,
    marginTop: 25,
    marginBottom: 25,
  },
  checkbox: {
    padding: 0,
    margin: 0,
    marginRight: 5,
  },
  rememberMe: {
    fontSize: '16px',
    lineHeight: '19px',
    color: '#494949',
  },
};
