// react-components/utils/api.js
import axios from 'axios';

import { HOST } from './constants';

export const Api = {
  login: function ({ email, password }) {

    const userCreds = {
      email: email,
      password: password
    };

    return axios.post(`${HOST}/login`, userCreds)
      .then(response => {
        return response;
      })
      .catch(error => {
        console.log(error);
        return error.response.data;
      });
  },
  register: function ({ email, password, password_confirmation, redirect }) {

    const user = {
      email: email,
      password: password,
      password_confirmation: password_confirmation,
      redirect: redirect,
    }

    return axios.post(`${HOST}/register`, user)
      .then(response => {
        return response;
      })
      .catch(error => {
        console.log("error: ", error.response);
        return error.response;
      });
  },
  passwordReset: function ({ email, redirect }) {

    const postData = {
      email: email,
      redirect: redirect,
    }

    return axios.post(`${HOST}/password/email`, postData)
      .then(response => {
        return response;
      })
      .catch(error => {
        return error.response;
      });
  },
  passwordUpdate: function ({ token, email, password, passwordConfirmation }) {

    const postData = {
      token: token,
      email: email,
      password: password,
      password_confirmation: passwordConfirmation
    }

    return axios.post(`${HOST}/password/reset`, postData)
      .then(response => {
        return response;
      })
      .catch(error => {
        return error.response;
      });
  },
}
