// react-components/utils/styles.js
export const styles = {
  componentContainer: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
    width: '450px',
  },
  container: {
    fontFamily: 'proxima-nova',
    boxShadow: '0 10px 40px rgba(0, 0, 0, .2)',
    width: '100%',
  },
  header: {
    fontSize: '22px',
    fontFamily: 'Arial',
    fontWeight: '300',
    lineHeight: '27px',
    color: '#494949',
    paddingBottom: 20,
  },
  form: {
    backgroundColor: '#FFFFFF',
    textAlign: 'left',
    width: '100%',
  },
  labelContainer: {
    marginBottom: '5px',
    marginTop: '10px',
  },
  label: {
    height: '19px',
    color: '#A5A5A5',
    fontFamily: 'proxima-nova',
    fontSize: '16px',
    fontWeight: 500,
    lineHeight: '19px',
  },
  input: {
    fontSize: 16,
    color: '#1B1B1B',
    fontFamily: 'proxima-nova',
    fontWeight: 300,
    lineHeight: '19px',
    border: '1px solid rgba(0,0,0,0.1)',
    height: '52px',
    backgroundColor: '#FFFFFF',
    borderRadius: '3px',
    padding: '15px',
    width: '100%',
    marginBottom: '10px',
  },
  button: {
    fontSize: 16,
    fontWeight: 700,
    height: 52,
  },
  error: {
    color: '#C00',
    fontFamily: 'proxima-nova',
    fontSize: '16px',
    lineHeight: '19px',
    fontWeight: 500,
    textAlign: 'left',
    margin: 0,
    padding: 0,
    marginBottom: 20,
  },
  link: {
    color: '#2D8EFF',
    fontSize: '16px',
    fontFamily: 'proxima-nova',
    fontWeight: 500,
    lineHeight: '19px',
    paddingTop: '15px',
    paddingBottom: '15px',
  },
  buttonLink: {
    color: '#2D8EFF',
    fontSize: '16px',
    fontFamily: 'proxima-nova',
    fontWeight: 500,
    lineHeight: '19px',
    paddingTop: '15px',
    paddingBottom: '15px',
    border: '0',
    cursor: 'pointer',
  },
  smallLink: {
    color: '#2D8EFF',
    fontFamily: 'proxima-nova',
    fontWeight: 500,
    lineHeight: '19px',
    paddingTop: '15px',
    paddingBottom: '15px',
    fontSize: '14px',
    margin: 0,
    padding: 0,
    marginTop: 10,
  },
  statement: {
    fontSize: '16px',
    lineHeight: '19px',
    color: '#494949',
  },
  bottomLine: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingTop: 20,
    width: '100%',
    height: '100px',
  },
  bottomLineOutsideContainer: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    width: '450px',
    marginTop: 10,
  },
  bottomLinks: {
    fontFamily: 'proxima-nova',
    fontSize: '14px',
    fontWeight: 500,
    lineHeight: '17px',
    color: '#494949',
  },
  button: {
    height: '45px',
    width: '100px',
    background: 'linear-gradient(90deg, #2D8EFF 0%, #39F 100%)',
    boxShadow: '0 2px 4px 0 rgba(0,0,0,0.2)',
    borderRadius: '3px',
    border: 'transparent',
    fontFamily: 'proxima-nova',
    color: '#FFFFFF',
    fontSize: '16px',
    fontWeight: 700,
    lineHeight: '16px',
    textAlign: 'center',
  },
  privacyLink: {
    marginRight: '15px',
    fontFamily: 'proxima-nova',
    fontSize: '14px',
    fontWeight: 500,
    lineHeight: '17px',
    color: '#494949',
  },
  termsLink: {
    fontFamily: 'proxima-nova',
    fontSize: '14px',
    fontWeight: 500,
    lineHeight: '17px',
    color: '#494949',
  },
  passwordContainer: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    margin: 0,
    padding: 0,
  },
  passwordStatement: {
    color: '#A5A5A5',
    fontFamily: 'proxima-nova',
    fontWeight: 500,
    textAlign: 'left',
    fontSize: '13px',
    lineHeight: '16px',
    margin: 0,
    padding: 0,
    marginBottom: 15,
  },
  leftItem: {
    width: '48%',
  },
  rightItem: {
    width: '48%',
  },
  invalidInput: {
    border: '1px solid #C00',
  },
  passwordStrengthDisplay: {
    fontSize: '16px',
    fontWeight: 500,
    lineHeight: '19px',
    color: '#A5A5A5',
    margin: 0,
    padding: 0,
    marginBottom: 20,
    width: '100%',
    textAlign: 'left',
  },
  mainContent: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
  },
  bold: {
    fontWeight: '600',
  },
  instructionsList: {
    color: '#444',
    fontFamily: 'proxima-nova',
    fontSize: '16px',
    lineHeight: '19px',
    textAlign: 'left',
    margin: 0,
    padding: 0,
  },
  qrCodeContainer: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    padding: 10,
    margin: 10,
    paddingBottom: 0,
    marginBottom: 0,
  },
  authInstructions: {
    padding: 0,
    margin: 0,
    textAlign: 'left',
    marginBottom: 20,
    width: '100%',
  },
  instructionStatement: {
    fontFamily: 'proxima-nova',
    fontWeight: 500,
    fontSize: '16px',
    lineHeight: '19px',
    color: '#444',
  },
  deepPanel: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    border: '1px solid #E5E5E5',
    backgroundColor: '#FFF',
    padding: 20,
    marginTop: 10,
    marginBottom: 10,
  },
  googleCode: {
    color: '#494949',
    fontFamily: 'proxima-nova',
    fontSize: '16px',
    fontWeight: '600',
    letterSpacing: '-0.25px',
    lineHeight: '19px',
    margin: 0,
    padding: 0,
  },
  section: {
    width: '100%',
  },
  radioInputs: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
  },
  radio: {
    border: '10px solid #2D8EFF',
    borderRadius: '10px',
    // backgroundColor: '#2D8EFF',
    color: '#2D8EFF',
    marginBottom: 10,
    width: 30,
    // '-webkit-appearance': 'none',
    // '-moz-appearance': 'none',
  },
  passwordListTitle: {
    color: '#494949',
    fontFamily: 'proxima-nova',
    fontSize: '14px',
    fontWeight: 500,
    letterSpacing: '-0.22px',
    lineHeight: '17px',
  },
  passwordSuccess: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center',
    color: '#55C400',
    fontFamily: 'proxima-nova',
    fontSize: '14px',
    fontWeight: 500,
    letterSpacing: '-0.22px',
    lineHeight: '17px',
  },
  passwordError: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center',
    color: '#C00',
    fontFamily: 'proxima-nova',
    fontSize: '14px',
    fontWeight: 500,
    letterSpacing: '-0.22px',
    lineHeight: '17px',
  },
  passwordErrorIcon: {
    color: '#C00',
    width: 12,
    height: 12,
    marginRight: 10,
  },
  passwordSuccessIcon: {
    color: '#55C400',
    width: 12,
    height: 12,
    marginRight: 10,
  },
  helpIcon: {
    width: 15,
    height: 15
  },
  logout: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    paddingTop: 20,
    width: '100%',
    height: '100px',
  },
};
