// react-components/src/Auth/index.js
import React, { Component } from 'react';

import Login from '../Login';
import Register from '../Register';
import AuthCode from '../AuthCode';
import PasswordReset from '../PasswordReset';
import PasswordUpdate from '../PasswordUpdate';
import AuthenticatorSetup from '../AuthenticatorSetup';
import AuthenticatorDisable from '../AuthenticatorDisable';
import { styles } from '../../utils/styles';

export default class Auth extends Component {

  constructor(props) {
    super(props);

    this.state = {
      display: 'login',
      initialDisplayComponent: 'login',
      showInitialDisplay: true,
    };

    this.displayComponent = this.displayComponent.bind(this);
    this.signInCB = this.signInCB.bind(this);
    this.authenticatorSetupCB = this.authenticatorSetupCB.bind(this);
    this.forgotPasswordCB = this.forgotPasswordCB.bind(this);
    this.onClickCancel = this.onClickCancel.bind(this);
    this.onClickDone = this.onClickDone.bind(this);
  }

  componentDidMount() {
    const { initialDisplayComponent } = this.props;

    this.setState({
      initialDisplayComponent: initialDisplayComponent,
      showInitialDisplay: true,
    });
  }

  displayComponent(componentName) {
    this.setState({
      display: componentName,
      showInitialDisplay: false,
    });
  }

  signInCB() {
    this.displayComponent('sign_in');
  }

  authenticatorSetupCB() {
    this.displayComponent('authenticator_setup');
  }

  forgotPasswordCB() {
    this.displayComponent('password_reset');
  }

  onClickCancel(toDisplay) {
    this.displayComponent(toDisplay);
  }

  onClickDone(toDisplay) {
    this.displayComponent(toDisplay);
  }

  render () {

    const { display, initialDisplayComponent, showInitialDisplay } = this.state;

    let componentToDisplay;
    let displayComponentValue;

    if (initialDisplayComponent && showInitialDisplay) {
      displayComponentValue = initialDisplayComponent;
    } else {
      displayComponentValue = display;
    }

    switch (displayComponentValue) {
      case 'sign_in':
        componentToDisplay = (
          <Login
            displayComponent={this.displayComponent}
          />
        );
      break;
      case 'sign_up':
        componentToDisplay = (
          <Register
            displayComponent={this.displayComponent}
            successCB={this.authenticatorSetupCB}
          />
        );
      break;
      case 'verification_code':
        componentToDisplay = (
          <AuthCode
            displayComponent={this.displayComponent}
          />
        );
      break;
      case 'password_reset':
        componentToDisplay = (
          <PasswordReset
            displayComponent={this.displayComponent}
            redirect={"https://tokenmarketplace.com"}
            signInCB={this.signInCB}
          />
        );
      break;
      case 'password_update':
        componentToDisplay = (
          <PasswordUpdate
            displayComponent={this.displayComponent}
          />
        );
      break;
      case 'authenticator_setup':
        componentToDisplay = (
          <AuthenticatorSetup
            displayComponent={this.displayComponent}
          />
        );
      break;
      case 'authenticator_disable':
        componentToDisplay = (
          <AuthenticatorDisable
            displayComponent={this.displayComponent}
            forgotPasswordCB={this.forgotPasswordCB}
            // onClickCancel={() => this.onClickCancel('login')}
            onClickDone={() => this.onClickDone('login')}
          />
        );
      break;
      default:
        componentToDisplay = (
          <Login
            displayComponent={this.displayComponent}
          />
        );
      break;
    }

    return (
      <div style={styles.componentContainer}>
        { componentToDisplay }
      </div>
    );
  }
}
