// react-components/src/AuthenticatorSetup/googleSetup.js
import React, { Component } from 'react';
import {CopyToClipboard} from 'react-copy-to-clipboard';

import { styles } from '../../utils/styles';

export default class GoogleSetup extends Component {

  constructor(props) {
    super(props);

    this.state = {
      googleCode: 'abcd efgh ijkl mnop qrstu vwxyz',
      copied: false,
    };

    this.onCopy = this.onCopy.bind(this);
  }

  componentDidMount() {
    this.setState({
      copied: false,
    });
  }

  onCopy() {
    this.setState({
      copied: true,
    });
  }

  render () {
    const { googleCode } = this.state;
    const { onClick } = this.props;

    return (
      <div style={styles.authInstructions}>
        <p style={styles.instructionStatement}>
          Instructions for Google Authenticator App
        </p>
        <ol style={styles.instructionsList}>
          <li>1. Tap the <span style={styles.bold}>plus symbol (+).</span></li>
          <li>2. Tap <span style={styles.bold}>Manual entry.</span></li>
          <li>3. Enter your email address and this key:
            <div style={styles.deepPanel}>
              <p style={styles.googleCode}>{googleCode}</p>

              <CopyToClipboard text={googleCode}
                onCopy={this.onCopy}>
                <span>
                  <a
                    style={{
                      ...styles.link,
                      fontSize: '14px',
                      margin: 0,
                      padding: 0,
                      marginTop: 10,
                    }}
                    href="#"
                  >
                    Copy to Clipboard
                  </a>
                </span>
              </CopyToClipboard>

            </div>
          </li>
          <li>4. Make sure that <span style={styles.bold}>Time based</span> is turned on, and tap
            <span style={styles.bold}>{" "}Add</span> to finish.
          </li>
        </ol>
        <br/>
        <a
          style={{
            ...styles.link
          }}
          href="#"
          onClick={onClick}
        >
          View instructions for Authy Authenticator
        </a>
      </div>
    );
  }
}
