// react-components/src/AuthenticatorSetup/verification.js
import React, { Component } from 'react';
import { SlideDown } from 'react-slidedown';
import 'react-slidedown/lib/slidedown.css';

import { styles } from '../../utils/styles';
import Container from '../Shared/Container';
import BottomLinks from '../Shared/BottomLinks';

export default class Verification extends Component {

  constructor(props) {
    super(props);

    this.state = {
      verification_code: '',
      error: false,
      errorMessage: '',
    };

    this.updateVerificationCode = this.updateVerificationCode.bind(this);
  }

  updateVerificationCode(event) {
    this.setState({
      verification_code: event.target.value,
    });
  }

  render () {
    const { verification_code, error, errorMessage } = this.state;
    const { onClickVerify, verificationCode, updateVerificationCode } = this.props;

    return (
      <div style={styles.authInstructions}>
        <p style={styles.instructionStatement}>
          Please enter the 6-digit verification code from your Authenticator app.
        </p>

        <div
          style={{
            ...styles.labelContainer,
          }}
        >
          <label
            style={{
              ...styles.label,
            }}
            htmlFor="verification_code"
          >
            Enter Code
          </label>
        </div>
        <input
          style={{
            ...styles.input,
          }}
          type="text"
          name="verification_code"
          placeholder="Verification Code"
          value={verificationCode}
          onChange={updateVerificationCode}
          required
        />
        <SlideDown
          closed={!error}
        >
          <p
            style={{
              ...styles.error,
            }}
          >
            { errorMessage }
          </p>
        </SlideDown>
      </div>
    );
  }
}
